### PR TITLE
Xarm testing

### PR DESCRIPTION
**Pull request by dxyang**
Creating a pull request so that I can add comments to this branch

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (dxyang) [inline .idea/.gitignore:1]: We should just add .idea to the .gitignore file at the root of the repo

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
+import numpy as np
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: maybe move this into a new folder called bc at the root of the repo? since bc isn't really related to reward extraction, it's just supervised learning

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+from torch.utils.data import Dataset, DataLoader
+from typing import Tuple
+from robot.data import RoboDemoDset
+from robot.base_robot import XArmTaskSpaceEnv
+# export PYTHONPATH="${PYTHONPATH}:/home/xarm/Documents/JacobAndDavin/test/rewardlearning-robot"
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
+class BCNet(nn.Module):
+    def __init__(self) -> None:
+        super(BCNet, self).__init__()
+        self.fc1 = nn.Linear(3, 30)
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: You can just use this instead of writing a MLP every time

https://github.com/dxyang/rewardlearning-robot/blob/main/reward_extraction/models.py#L20

```
from reward_extraction.models import MLP

self.net = MLP(3, 300, 3, 3, nn.Tanh())
```

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        self.relu = nn.LeakyReLU()
+    
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.relu(x)    
+        x = self.fc2(x)
+        x = self.relu(x)
+        x = self.fc4(x)
+        x = self.relu(x)
+        x = self.fc3(x)
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: you probably wanna tanh the output so the range is squished to become 0 and 1.

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+        return x
+    
+
+def data(SCALE) -> Tuple[DataLoader, DataLoader]:
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: what's the point of scaling things?

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    state = []
+    targets = []
+    for traj in expert_data:
+        joint_poses = traj['eef_pose']
+        # delta_targets = [] # np.zeros((len(joint_poses)-1, 3))
+        for i in range(0, len(joint_poses) -1):
+            if np.sum(joint_poses[i+1] - joint_poses[i]) == 0:
+                continue
+            state.append(joint_poses[i] / SCALE)
+            targets.append(joint_poses[i+1] / SCALE - joint_poses[i] / SCALE)
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: could be simplfied
```
targets = joint_poses[: -1] - joint_poses[1:] # these I assume are just deltas
states = joint_poses[:-1]
```

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    perm = torch.randperm(len(state))
+    state = torch.Tensor(state)
+    targets = torch.Tensor(targets)
+    state = state[perm]
+    targets = targets[perm]
+    test_idx = int(len(state) * 0.8)
+    train_x = state[:test_idx]
+    train_y = targets[:test_idx]
+    test_x = state[test_idx:]
+    test_y = targets[test_idx:]
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: https://pytorch.org/docs/stable/data.html?highlight=random_split#torch.utils.data.random_split

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        # super().__init__()
+        self.x = x
+        self.y = y
+    def __len__(self):
+        return(len(self.x))
+
+    def __getitem__(self, idx):
+        return self.x[idx], self.y[idx]
+
+def one_train_iter(net, optimizer, train_loader):
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: nit but this would be a full epoch of your train data instead of just an iteration

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        preds = net(data)
+        loss = critirion(preds, labels)
+        losses.append(loss.item())
+        loss.backward()
+        optimizer.step()
+    print(f'Average train loss is {np.mean(losses)}')
+    return (np.mean(losses))
+
+def eval(net, test_loader):
+    critirion = nn.MSELoss()
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: doesn't really model for the simple MLP you have above, but you should call model.eval() to put the model in eval mode in case it has something like batch normalization or some sort of distribution sampling

File: reward_extraction/behavior_cloning/behavior_cloning.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+def try_until_estop(model, SCALE):
+    env = XArmTaskSpaceEnv(
+        control_frequency_hz=10,
+        use_camera=True,
+        # TODO: create some sort of button pressing mechanism to open and close the gripper,
+        random_reset_home_pose=False
+    )
+    obs = env.reset()
+    model.eval()
+    model.double()
```
Comment (dxyang) [inline reward_extraction/behavior_cloning/behavior_cloning.py]: generally leaving models with float precision is fine instead of doubles

File: robot/base_robot.py [hunk 11]
```diff
@@ -151,6 +156,200 @@ def _calculate_reward(self, obs) -> float:
         return 0
 # test commit since github is weird
 
+
+class XArmTaskSpaceEnv(RobotEnv):
```
Comment (dxyang) [inline robot/base_robot.py]: there's a lot of copy and paste in this class. you can use inheritance to your advantage here to clean this up (i.e., override the method, do what you want that's custom, and then call the parent classes version of that function)

then you don't need to redundantly code up functions that exist already also (ex: action_space, _calculate_reward)

you probably want to do somethng like:
```
class XArmEnv(TaskSpaceRobotEnv):
  def __init__(self, xarm_ip: str = '192.168.1.220', **kwargs):
     self.xarm_ip = xarm_ip
     super().__init__(self, **kwargs)
```

Comment (dxyang) [reply robot/base_robot.py]: also this class should probably live in the `xarm_env.py` file so you shouldn't be making any changes to `base_robot.py`

File: robot/data.py [hunk 12]
```diff
@@ -58,10 +58,10 @@ def add_traj(self, rgbs, joint_angles, eef_poses):
...
         # (bs, traj_length, joint_angles)
         # (bs, eef_poses)
-        if self.rgb_shape is None:
-            self.rgb_shape = rgbs.shape[1:]
-            self.ja_shape = joint_angles.shape[1:]
-            self.eef_shape = eef_poses.shape[1:]
+        self.rgb_shape = rgbs.shape[1:]
+        self.ja_shape = joint_angles.shape[1:]
+        self.eef_shape = eef_poses.shape[1:]
+
```
Comment (dxyang) [inline robot/data.py:64]: i don't think this changes any functionality? if shapes are changing as you're adding things, something else is wrong

File: robot/xarm_env.py [hunk 13]
```diff
@@ -1,8 +1,243 @@
...
+from xarm.wrapper import XArmAPI
+from reward_extraction.reward_functions import RobotLearnedRewardFunction
+from robot.utils import Rate
+from r3m import load_r3m
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 
-'''
-TODO: define
-'''
+class XArmEnv(RobotEnv):
```
Comment (dxyang) [inline robot/xarm_env.py]: same comments as the xarmenv above? seems like you might've copy/paste moved things and not cleaned up

File: robot/xarm_env.py [hunk 14]
```diff
@@ -1,8 +1,243 @@
...
+        self.hz = control_frequency_hz
+        self.rate = Rate(control_frequency_hz)
+        self.control_mode = control_mode
+        self.use_gripper = use_gripper
+        self.random_reset_home_pos = random_reset_home_pose
+        self.xarm_ip = xarm_ip
+        self.robot = None
+        # TODO: This is bad, we should have a better home pos
+        if self.control_mode == 'default':
+            self.home_xyz = [553,29,435]
```
Comment (dxyang) [inline robot/xarm_env.py]: you should get in the habit of adding units (radians, degrees, centimeters, meters) for robot pose related things. people will generally assume meters, but this is clearly not meters. things also start to get messy say if you were using realsense depth which gets returned in mm.

**Review by dxyang — COMMENTED** [review]
some comments from a quick passthrough

File: .idea/.gitignore (hunk 1 above)
Comment (Davin05) [reply .idea/.gitignore:1]: sounds good, just did that, thanks

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 2 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: That makes sense, we weren't sure where to create a new folder so we just temporarily put it in rewad_extraction/. Just moved it to the root, thanks

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 3 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: oh that's nice, I will look into that module. Just update BCNet to MLP.

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 10 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: we ran into error when we left it at float: `expected scaler type double` but found float. Maybe we made an error somewhere I will look into it thanks

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 5 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: we ran into loss = Nan problem (likely) due to gradient explosion, so we were wondering if we scale it by a hundredth when we train and the scale it back when we deploy would resolve the issue. But then we got it to work without scaling , so i think it is unnecessary, will remove after some testing making sure there is no weird bug left unattended. Thanks for pointing it out

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 6 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: will do the simplification after some testing, thanks

File: robot/base_robot.py (hunk 11 above)
Comment (Davin05) [reply robot/base_robot.py]: oh yeah we overlook this part, we will refactor the code a little bit and keep this in mind in the future

File: robot/xarm_env.py (hunk 13 above)
Comment (Davin05) [reply robot/xarm_env.py]: sounds good, deleted

File: robot/xarm_env.py (hunk 14 above)
Comment (Davin05) [reply robot/xarm_env.py]: unit added

File: reward_extraction/behavior_cloning/behavior_cloning.py (hunk 4 above)
Comment (Davin05) [reply reward_extraction/behavior_cloning/behavior_cloning.py]: pinned you on Slack for this issue
